### ISSUE TITLE
Why not let Bazel compile C++ as well?

### ISSUE THREAD

**Issue by 1e100** [grailbio/rules_r#57]
Currently, `r_pkg` leans pretty heavily on R to build *.cpp code, which is pretty slow. 

What if, instead, C++ code was built with Bazel (which is fast), and then passed to `r_pkg` through `cc_dep`, already prebuilt? I've hacked up something like this and it seems to work fine, and it's much faster. I had to use an empty *.cpp file to get `r_pkg` to link the *.so, and I had to put libRcpp.so into a location where it could easily be found, but other than that this required no changes to the C++ code. The resulting package seems to work OK under `r_unit_test`. Not sure what `r_pkg_test` would think of it, but it'll probably be fine as well, since it works OK with other `cc_deps` and this is no different.

So basically the suggestion here is to modify the rule such that it compiles all C++ source with `cc_library` (and lines up the *.so's and headers for that to work, so symlinking hack is not necessary), and then feeds the result into `build.sh` much like it would feed `cc_deps`.

**Comment by siddharthab** [comment]
For packages managed by the user, they can define their cc_library rules, use cc_deps, and manage which files from the 'src' directory get included in r_pkg. That's why I did not make this an automated feature, because advanced users can write their BUILD files like that.

It gets tricky when we want to build external packages through auto-generated BUILD files. The general problem of translating Makevars into a cc_library target is hard. Another thing we will have to think is that running the configure script and then parsing the Makevars file will need to be run at repository fetch time and not at build time, which could be OK, but will need an interface change to r_pkg and r_repository.

One way in which I can help people is to give them a way to extract the .so and header files from other R packages like Rcpp, so they can depend on those in their cc_library target.

And maybe if the package does not have a Makevars file or a configure script, then I can use bazel's starlark C API automatically instead of expecting the user to use cc_deps in their BUILD file.

This will take considerable effort and I am not sure how much ground it will cover, so this issue will be a low priority.

**Comment by siddharthab** [comment]
Just realised that one of the main reasons you mention in the issue is speed. At GRAIL, we set the `MAKE` environment variable in the R bazel toolchain to be something like `make -j 8`. This makes R compile all the cc code with parallelism as a hard-coded value of 8, and gives us reasonable speed for most packages.

**Comment by 1e100** [comment]
Yes, we use `-j4` as well, but that's still nowhere near as fast as Bazel itself, due to the extensive caching that it employs. Furthermore, we use Bazel's remote cache feature as well. Make can't take advantage of that either. I hear what you're saying regarding makevars and so on - this is not something one could just enable on existing packages in the general case. Indeed it might not even be appropriate for `rules_r` to support something like this.

That said, I think there are certain scenarios, such as our own, where this could be very useful even if the resulting packages aren't necessarily CRAN-kosher.

We basically use Bazel to build tools which we then deploy into a workflow on AWS. Those tools use quite a few R packages, which `rules_r` currently builds from us. A number of large packages we're working on depends on Cytolib, which we've modified quite extensively, and where much of the dev work is happening. 

Every time we modify something in Cytolib, things have to be rebuilt (in continuous integration and on dev machines) and that takes a while. Hence the idea of letting Bazel do fast, incremental builds, and provide the build products to rules_r for linking and inclusion into packages.

Another issue is that anything built by Make does not take advantage of Bazel's command line settings (with the exception of coverage which you implemented specifically). If, for instance, I would like to instrument _all_ native code for detecting memory leaks, thread safety, or for sample-based profiling, there's no straightforward way to do so. For Bazel-built code this is trivial - you just add a few configs to .bazelrc and it works throughout.

The main problem with implementing this is actually getting C++ deps from other R packages s.t. they could be separately compiled using a `cc_library` rule. If that can be done, then the rest of the implementation becomes trivial - one just needs to get the C++ deps, shove them into `cc_library` rule, and out pops an object file that Make could link into an `*.so` in basically no time at all.

But as you know, such things require time, experimentation, and a lot of head scratching, so I thought you might have some insight into whether this is feasible.

**Comment by siddharthab** [comment]
I understand your pain. This is something to which I have also given some thought, but there is no clear solution. At GRAIL, we don't use Rcpp or other "Linking To" type dependencies that much, so a `cc_import` rule for an R package was never a big requirement for us. But as I said in my first comment above, this is something where I think I can help. Let me try something and get back to you.

I hope I am understanding your main problem above correctly.

**Comment by siddharthab** [comment]
@1e100 , I have created a pull request that needs a bit more work, but puts the foundations in place.

I have written a very basic test in `//rcpp:test`, but I am currently getting wrong linker flags on macOS, and a segfault on linux. I would appreciate if you are happy to debug and take over from here. Also feel free to restructure to suit your needs better.

I am done with the rest of my work on this project for the time being, so won't be able to give more time to it in the near future.

**Comment by 1e100** [comment]
Much appreciated! Let me take a look and get back to you. I hope we can make this work.

**Comment by siddharthab** [comment]
I was actually able to work on it a little more and get a working test going. Please see changes in tests/exampleC in the pull request.

**Comment by siddharthab** [comment]
Release 0.6.1 should now have the best I can do for this.

Coverage is still missing from sources in cc_deps, which can be worked on at a later point.
